src/range.rs

```diff
@@ -1,98 +1,259 @@
-            pub end: $addr_type,
-        }
+/// A range of a given memory address type `A`.
+///
+/// The range is inclusive on the start and exclusive on the end.
+/// It is empty if `start >= end`.
+///
+/// # Example
+///
+/// ```
+/// use memory_addr::{AddrRange, VirtAddr};
+///
+/// let range = AddrRange::<VirtAddr>::new(0x1000.into(), 0x2000.into());
+/// assert_eq!(range.start, 0x1000.into());
+/// assert_eq!(range.end, 0x2000.into());
+/// ```
+#[derive(Clone, Copy)]
+pub struct AddrRange<A: MemoryAddr> {
+    /// The lower bound of the range (inclusive).
+    pub start: A,
+    /// The upper bound of the range (exclusive).
+    pub end: A,
+}
 
-        impl $name {
-            /// Creates a new address range.
-            #[inline]
-            pub const fn new(start: $addr_type, end: $addr_type) -> Self {
-                Self { start, end }
-            }
-
-            /// Creates a new address range from the start address and the size.
-            #[inline]
-            pub const fn from_start_size(start: $addr_type, size: usize) -> Self {
-                Self {
-                    start,
-                    end: <$addr_type>::from(usize!(start) + size),
-                }
-            }
-
-            /// Returns `true` if the range is empty (`start >= end`).
-            #[inline]
-            pub const fn is_empty(self) -> bool {
-                usize!(self.start) >= usize!(self.end)
-            }
-
-            /// Returns the size of the range.
-            #[inline]
-            pub const fn size(self) -> usize {
-                usize!(self.end) - usize!(self.start)
-            }
-
-            /// Checks if the range contains the given address.
-            #[inline]
-            pub const fn contains(self, addr: $addr_type) -> bool {
-                usize!(self.start) <= usize!(addr) && usize!(addr) < usize!(self.end)
-            }
-
-            /// Checks if the range contains the given address range.
-            #[inline]
-            pub const fn contains_range(self, other: Self) -> bool {
-                usize!(self.start) <= usize!(other.start) && usize!(other.end) <= usize!(self.end)
-            }
-
-            /// Checks if the range is contained in the given address range.
-            #[inline]
-            pub const fn contained_in(self, other: Self) -> bool {
-                other.contains_range(self)
-            }
-
-            /// Checks if the range overlaps with the given address range.
-            #[inline]
-            pub const fn overlaps(self, other: Self) -> bool {
-                usize!(self.start) < usize!(other.end) && usize!(other.start) < usize!(self.end)
-            }
-        }
+/// Methods for [`AddrRange`].
+impl<A> AddrRange<A>
+where
+    A: MemoryAddr,
+{
+    /// Creates a new address range.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use memory_addr::VirtAddrRange;
+    ///
+    /// let range = VirtAddrRange::new(0x1000.into(), 0x2000.into());
```

Comment:
Is it still appropriate to use `VirtAddrRange` here?

Reply:
It's better to show `AddrRange<usize>` in the example.